segmentstore/server/src/test/java/io/pravega/segmentstore/server/containers/StreamSegmentMetadataTests.java

```diff
@@ -212,6 +218,47 @@ private void checkAttributesEqual(Map<UUID, Long> expected, Map<UUID, Long> actu
         }
     }
 
+    /**
+     * Validates that we can safely iterate over a StreamSegmentMetadata.AttributesView elements while the backing
+     * StreamSegmentMetadata maps are being modified.
+     */
+    @Test
+    public void testAttributeConcurrentOperations() throws ExecutionException, InterruptedException, TimeoutException {
+        StreamSegmentMetadata metadata = new StreamSegmentMetadata(SEGMENT_NAME, SEGMENT_ID, CONTAINER_ID);
+
+        // Step 1: initial set of attributes.
+        Random rnd = new Random(0);
+        val expectedAttributes = generateAttributes(rnd);
+
+        metadata.updateAttributes(expectedAttributes);
+        ExecutorService executorService = null;
+        CompletableFuture<Void> iterateOverView;
+        CompletableFuture<Void> modifyBackingAttributes;
+        try {
+            executorService = ExecutorServiceHelpers.newScheduledThreadPool(4, "testAttributes");
+            Map<UUID, Long> attributesView = metadata.getAttributes();
+            iterateOverView = CompletableFuture.runAsync(() -> {
+                for (int i = 0; i < 100; i++) {
+                    new HashMap<>(attributesView); // Internally, this iterates over the input map.
+                }
+            }, executorService);
+            modifyBackingAttributes = CompletableFuture.runAsync(() -> {
+                for (int i = 0; i < 100; i++) {
+                    metadata.updateAttributes(generateAttributes(new Random(i)));
```

Comment:
I see what you're trying to do there, but the thing is you are not really guaranteed to test your change at all. You may or may not have a concurrent iteration with a modification. 

This is a non-deterministic way of verifying things. May I suggest a simpler approach:
1. Create your metadata and put some attributes in it.
2. Get an iterator of those attributes (and maybe fetch 1 item)
3. Modify the attributes
4. Attempt to get the remainder of those attributes from the iterator.

You'll see that the "old code" would throw the `ConcurrentModificationException` while the new one should not. The advantage of this test is that it's deterministic and it explicitly verifies what you are changing.

Reply:
Right, I changed the test according to your suggestion.